File: solutions/csharp/grade-school/1/GradeSchool.cs
using StudentsRoster = System.Collections.Generic.SortedSet<string>;

public class GradeSchool
{
    private readonly SortedDictionary<int, StudentsRoster> _studentsByGrade = [];
    private readonly StudentsRoster _roster = new();

    public bool Add(string student, int grade)
    {
        if (!_roster.Add(student))
        {
            return false;
        }

        UpdateGrade(grade, student);
        return true;
    }

    public IEnumerable<string> Roster() => _studentsByGrade.Values
        .SelectMany(students => students);

    public IEnumerable<string> Grade(int grade) => StudentsForGrade(grade);

    private StudentsRoster StudentsForGrade(int grade) =>
        _studentsByGrade.TryGetValue(grade, out var students) ? students : EmptyRoster();

    private void UpdateGrade(int grade, string student)
    {
        var students = StudentsForGrade(grade);
        students.Add(student);
        _studentsByGrade[grade] = students;
    }

    private static StudentsRoster EmptyRoster() => [];
}

File: solutions/kotlin/grade-school/1/src/main/kotlin/GradeSchool.kt
class School {
    private val studentsByGrade = mutableMapOf<Int, List<String>>()

    fun add(student: String, grade: Int) {
        require(student !in roster())

        studentsByGrade[grade] = studentsIn(grade) + student
    }

    fun grade(grade: Int): List<String> = studentsIn(grade).sorted()

    fun roster(): List<String> = studentsByGrade
        .toSortedMap()
        .values
        .flatMap { students -> students.sorted() }

    private fun studentsIn(grade: Int): List<String> = studentsByGrade.getOrDefault(grade, emptyList())
}
